### PR TITLE
Add fanout for full-graph inference

### PR DESCRIPTION
**Pull request by isratnisa**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/inference/ep_infer.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ class GSgnnEdgePredictionInfer(GSInfer):
         The rank.
     """
 
-    def infer(self, loader, save_embed_path, save_prediction_path=None,
+    def infer(self, loader, fanout, save_embed_path, save_prediction_path=None,
```
Comment (zheng-da) [inline python/graphstorm/inference/ep_infer.py:43]: it should be called eval_fanout

File: python/graphstorm/model/gnn.py [hunk 2]
```diff
@@ -602,7 +602,8 @@ def loss_func(self):
         """
         return self._loss_fn
 
-def do_full_graph_inference(model, data, batch_size=1024, edge_mask=None, task_tracker=None):
+def do_full_graph_inference(model, data, batch_size=1024, fanout=-1, edge_mask=None,
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:605]: fanout is a list. its default value shouldn't be -1. probably None is a better value.

File: python/graphstorm/inference/ep_infer.py [hunk 3]
```diff
@@ -53,6 +53,8 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
         ----------
         loader : GSEdgeDataLoader
             The mini-batch sampler for edge prediction task.
+        fanout: list of int
```
Comment (classicsong) [inline python/graphstorm/inference/ep_infer.py:56]:
```suggestion
        eval_fanout: list of int
```

File: python/graphstorm/model/gnn.py [hunk 4]
```diff
@@ -602,7 +602,8 @@ def loss_func(self):
         """
         return self._loss_fn
 
-def do_full_graph_inference(model, data, batch_size=1024, edge_mask=None, task_tracker=None):
+def do_full_graph_inference(model, data, batch_size=1024, fanout=None, edge_mask=None,
```
Comment (classicsong) [inline python/graphstorm/model/gnn.py:605]: Please add some unit tests for fanout=None and fanout is not None.

File: python/graphstorm/model/gnn_encoder_base.py [hunk 5]
```diff
@@ -118,7 +118,7 @@ def dist_inference(g, gnn_encoder, get_input_embeds, batch_size, fanout,
                                                 partition_book=g.get_partition_book(),
                                                 ntype=ntype, force_even=False)
             # need to provide the fanout as a list, the number of layers is one obviously here
-            sampler = dgl.dataloading.MultiLayerNeighborSampler([fanout], mask=edge_mask)
+            sampler = dgl.dataloading.MultiLayerNeighborSampler([fanout[i]], mask=edge_mask)
```
Comment (classicsong) [inline python/graphstorm/model/gnn_encoder_base.py:121]: You need to check if fanout is None, set it to -1 otherwise use fanout[i].

File: python/graphstorm/inference/ep_infer.py (hunk 1 above)
Comment (isratnisa) [reply python/graphstorm/inference/ep_infer.py]: fixed.

File: python/graphstorm/model/gnn.py (hunk 2 above)
Comment (isratnisa) [reply python/graphstorm/model/gnn.py]: Done.

File: python/graphstorm/run/gsgnn_np/ml_nc_infer_ogbn_mag.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+---
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_np/ml_nc_infer_ogbn_mag.yaml:1]: Why we need to add this yaml?

File: python/graphstorm/trainer/ep_trainer.py [hunk 7]
```diff
@@ -162,7 +162,7 @@ def fit(self, train_loader, num_epochs,
                 if self.evaluator is not None and \
                     self.evaluator.do_eval(total_steps, epoch_end=False):
                     val_score = self.eval(model.module, val_loader, test_loader,
```
Comment (classicsong) [inline python/graphstorm/trainer/ep_trainer.py:164]: Did you change the interface of eval?

File: python/graphstorm/run/gsgnn_np/ml_nc_infer_ogbn_mag.yaml (hunk 6 above)
Comment (isratnisa) [reply python/graphstorm/run/gsgnn_np/ml_nc_infer_ogbn_mag.yaml]: Added by mistake. Removed.

File: python/graphstorm/dataloading/dataloading.py [hunk 8]
```diff
@@ -474,15 +475,18 @@ class GSgnnLinkPredictionTestDataLoader():
...
         The number of negative edges per positive edge
     """
-    def __init__(self, dataset, target_idx, batch_size, num_negative_edges):
+    def __init__(self, dataset, target_idx, batch_size, eval_fanout, num_negative_edges):
         self._data = dataset
         for etype in target_idx:
             assert etype in dataset.g.canonical_etypes, \
                     "edge type {} does not exist in the graph".format(etype)
         self._batch_size = batch_size
+        self._eval_fanout = eval_fanout
```
Comment (zheng-da) [inline python/graphstorm/dataloading/dataloading.py:489]: is this used?

File: python/graphstorm/dataloading/dataloading.py [hunk 9]
```diff
@@ -474,15 +475,18 @@ class GSgnnLinkPredictionTestDataLoader():
         The target edges for prediction
     batch_size: int
         Batch size
+    eval_fanout : list of int
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/dataloading/dataloading.py:478]: why do you need to modify this dataloader?

File: python/graphstorm/trainer/ep_trainer.py [hunk 10]
```diff
@@ -255,7 +255,8 @@ def eval(self, model, val_loader, test_loader, use_mini_batch_infer, total_steps
             test_pred, test_label = edge_mini_batch_gnn_predict(model, test_loader,
                                                                 return_label=True)
         else:
-            emb = do_full_graph_inference(model, val_loader.data, task_tracker=self.task_tracker)
+            emb = do_full_graph_inference(model, val_loader.data, fanout=val_loader._fanout,
```
Comment (zheng-da) [inline python/graphstorm/trainer/ep_trainer.py:258]: this code is not right

File: python/graphstorm/inference/ep_infer.py [hunk 11]
```diff
@@ -40,7 +40,7 @@ class GSgnnEdgePredictionInfer(GSInfer):
         The rank.
     """
 
-    def infer(self, loader, save_embed_path, save_prediction_path=None,
+    def infer(self, loader, eval_fanout, save_embed_path, save_prediction_path=None,
```
Comment (zheng-da) [inline python/graphstorm/inference/ep_infer.py:43]: if the data loader has fanout, why not just get the fanout from the data loader?

File: python/graphstorm/inference/ep_infer.py [hunk 12]
```diff
@@ -69,7 +69,7 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
         do_eval = self.evaluator is not None
         sys_tracker.check('start inferencing')
         self._model.eval()
-        embs = do_full_graph_inference(self._model, loader.data,
+        embs = do_full_graph_inference(self._model, loader.data, fanout=loader._fanout,
```
Comment (zheng-da) [inline python/graphstorm/inference/ep_infer.py:72]: you still need to expose `fanout` as a property

Comment (isratnisa) [reply python/graphstorm/inference/ep_infer.py]: Done.

File: python/graphstorm/inference/ep_infer.py (hunk 11 above)
Comment (isratnisa) [reply python/graphstorm/inference/ep_infer.py]: Done.

File: python/graphstorm/dataloading/dataloading.py (hunk 8 above)
Comment (isratnisa) [reply python/graphstorm/dataloading/dataloading.py]: Removed.

**Review by zheng-da — APPROVED** [review]
